### DISCUSSION TITLE
Remove padding in string keywords and align comments?

### DISCUSSION THREAD

**Discussion by wohler** [nom-tam-fits/nom-tam-fits#237] [Q&A] [answered]
Attila, our paths cross again!

I've received a report that the keywords that nom.tam.fits is emitting are less readable and may break our software. For example, nom.tam.fits wrote out the following keyword:
```
OBSERVAT= 'SOFIA   ' / Stratospheric Observatory for Infrared Astronomy
```
I've been asked to remove the spaces (padding) from short string values and line up the longer comments with the others as follows:
```
OBSERVAT= 'SOFIA'                   / Stratospheric Observatory for Infrared Astronomy
```
Is this possible?

**Comment by attipaci** [comment, marked as answer]
Hi Bill,

Well, that's tricky. Earlier versions of the FITS standard required that a closing quote may not come before character 21 (counting from 1) in the 80-character header line, and that shorter strings must be padded with empty spaces. FITS also explicitly specifies that any trailing spaces in strings ought to be ignored, no matter (but leading spaces are to be preserved).

As such, the aim of the library is to provide the best possible compatibility, including older tools written for earlier FITS standards in mind -- which means we want to keep the short string value padding. But that should not cause problems as long as the FITS requirement of ignoring trailing spaces is properly implemented with the library you use for reading. It seems to me that the fix should be happening on the reading side. Either the library you use needs to be fixed to drop the trailing spaces, as per FITS definition. Or, if that is not an option, you may want to patch your downstream code to remove the trailing spaces explicitly after parsing header strings...

If it's __astropy__ / __pyfits__ that you are using, you should probably file a bug report with them. Refer them to Section 4.2.1.1 of the [FITS 4.0 standard](https://fits.gsfc.nasa.gov/standard40/fits_standard40aa-le.pdf).
 
As for the alignment of the comments, this library aligns to a fixed position whenever that is possible, depending on the length of the value and/or comment. FITS really does not specify anything on comment alignment, so it's not something that a library should require for reading. This library adds some reasonable comment alignment to make headers visually pleasing to read, while trying to preserve as much of the comment as possible. If the alignment is causing issues with the reading software, that again should be a bug report for the library used for reading, as I don't believe anything this library does is incorrect or wrong.

**Comment by attipaci** [comment]
Note also, that your example of:

```
OBSERVAT= 'SOFIA'                   / Stratospheric Observatory for Infrared Astronomy
```
is longer than 80 characters allowed in the FITS header. Which is why the library aligns to the earlier mark to make the full comment fit. If the comment were shorter, s.t. it would fit into the 80-character space, the alignment would be what you expect above, e.g.:

```
OBSERVAT= 'SOFIA'                   / Stratospheric Obs. for IR Astronomy
```

**Reply by wohler** [reply]
That's a typo then. The comment should start in column 31, which puts the end of the comment in column 80. Thus, there wasn't a reason to start it in an earlier column. 

The alignment is for the benefit of human readers.

Thanks for your answer. I'll pass it along.

**Reply by attipaci** [reply]
PS. Also, I think the latest release (1.16.0) of the library improves on the consistency of the comment alignment over what it was in prior releases (1.15.2 and earlier). There is no `OBSERVAT` keyword in the SOFIA headers, but there is a `TELESCOP` keyword, which with 1.16.0 appears in the reduced HAWC+ FITS images as:

```
TELESCOP= 'SOFIA 2.5m'         / observatory name.
```

which is followed by the following keywords:

```
TELCONF = 'NASMYTH '           / telescope configuration.
TELRA   =             1.284636 / (hour) Boresight RA.
TELDEC  =    7.472117000000001 / (deg) Boresight DEC.
TELEQUI =                 2.E3 / Boresight epoch.
TELVPA  =            15.910983 / (deg) Boresight position angle.
```

As you can see they all align consistently...

**Comment by wohler** [comment]
Darn, I'm already using 1.16.0.

Here's an example that shows how our header is getting aligned. I agree that it shouldn't break readers, but it is harder for human readers to skim.
```
SIMPLE  =                    T / Java FITS: Wed Jan 12 08:53:39 PST 2022        
BITPIX  = 16 / bits per pixel (constant 16 for current imagers)                 
NAXIS   =                    2 / number of axes                                 
NAXIS1  =                 1024 / image width                                    
NAXIS2  =                 1024 / image height                                   
ORIGIN  = 'SOFIA   ' / Stratospheric Observatory for Infrared Astronomy         
DATE    = '2022-01-12T16:53:39.867' / date file created                         
HISTORY  Extracted from ARK file: fpi_node_b.ta_fpi.191203212042.ark            
FILENAME= 'FPI.20191203T212041.829Z.fits' / file name                           
OBSERVAT= 'SOFIA   ' / Stratospheric Observatory for Infrared Astronomy         
TELESCOP= 'SOFIA   '           / 2.5 meter telescope                            
IMAGTYP = 'CCD_IMAGE' / image type ('CCD_IMAGE', 'DARKFIELD' or 'FLATFIELD')    
INSTRUME= 'FPI     '           / 'FPI', 'FFI' or 'WFI'                          
DATE-OBS= '2019-12-03'         / date of exposure start                         
UTCSTART= '21:20:41.829Z'      / UTC of exposure start                          
UTCEND  = '21:20:42.829Z'      / UTC of exposure finish                         
EXPTIME =                  1.0 / [s]                                            
BINNING = '1x1     '           / pixel grouping (='1x1', '2x2' or '4x4')        
PIXBITS =                   14 / ADC bits (=8, 14, or 16)                       
DATAMAX =                16383 / detector full well maximum (=2^PIXBITS - 1)    
DATAMIN =                    0 / smallest value from detector                   
BZERO   =                 0.E0 / zero point to adjust pixel values              
IMAGE_SN=                30528 / unique image identifier                        
IMAGUSE = 'NORMAL  ' / image use (NORMAL, MONITOR, DARK_ACQUISITION, TEST)      
FILTPOS = 'FILTER_ND1'         / filter position                                
FOCLEN  = 5154.85 / [mm] focal length of camera optics (100 to 100000)          
FPIFOCUS= -32.19 / [mm] FPI focus (-300 to 300, or -1000 if unknown)            
FPIRETCL= -1.0 / FPI reticle brightness (0.0 to 1.0, or -1 if undef)            
END    
```
The example for this header in our ICD aligns the comments in column 27. This aligns all but four of the comments, whose values exceed column 27. None of the comments begin before column 27.

While it might be fun to maximize the number of aligned comments, a quick and dirty solution would be to add a setTargetCommentColumn() method for those who care. Comments for long values would be pushed right, or comments for short values and long comments would be pushed left.

**Reply by attipaci** [reply]
That is a workable solution. However, since it will affect the API, any such change will be slated for 1.17 earliest, which will not happen for a while. I'm certainly not planning on it before Summer...

Also, as you mention, moving the alignment mark to 27 won't be a huge change, as you go from 7 to 4 unaligned comments in the above example. All comments will never be aligned. So is it really such a big deal that you have a handful fewer comments unaligned? To me that does not seem like a huge improvement... (But, that's just my personal opinion.) Is that enough to justify increasing the API complexity with a corner-case method?

**Comment by attipaci** [comment]
PS. However, I can probably change the hard-coded default alignment to 27 for 1.16.1, slated for release in mid-March. That does not affect the API, and I really don't think it matters what the default is, as long as it is something reasonable. 

I would even argue that it could be even smaller, such as 24, which leaves 14 characters for the value of an aligned comment (vs 20 currently). 14 is sufficient to represent all practical integers and 32-bit floating point values, as well as strings up to 12 characters. 64-bit floating point values will typically require up to 18 value characters (hence align to 28) if no exponential form is needed (0.1 to 1e16), or up to 24 characters (align to 34) in exponential form. As such a practical default alignment point should be one of 24, 28, or 34... (27 is none of the above -- and neither is 30 to be fair...).

And beyond changing the default, I can still make it configurable starting 1.17...

**Comment by attipaci** [comment]
Created issue #283 and will make the required changes to allow configuring of the header alignment position in 1.17.0 (slated for release arounf Jun 15)... 

(Also created amnd closed issue #284 for changing how strings are padded/aligned -- with padding added as necessary before the opening quote (rather than before the closing quote), and string aligning to the right like numerical values... However, the FITS standard is quite explicit that string values must have their opening quote at byte 11 [counted from 1], and so padding before, or right-aligning is not an option at all...)

Code changes to follow soon...

**Reply by rmathar** [reply]
Noteworthy: the FITS standard has no end-of-line characters at all, but only implicit 80-byte cuts, so the entire 2880-blocks in the headers represent themselves in a natural way (in some generic editor) as a single long line with none of the vertical alignment discussed here...

**Comment by attipaci** [comment]
1.17.0 got released today, including the changes to header comment alignment!

**Reply by wohler** [reply]
Very cool, thanks! Unfortunately, SOFIA got canceled so I will not be able to pass on this goodness there, but I hope to make use of this in a distant galaxy at a distant time in the future...
